File: apps/web/lib/claimAuthorization.ts
// CD3-D — Central Public Claim Authorization. TypeScript mirror of
// app/services/claim_authorization.py (kept in sync by inspection, same
// convention as measurementSemantics.ts/claimProvenance's relationship to
// their backend counterparts).
//
// Authorization itself is a BACKEND concern (see the Python module's own
// docstring: enforcement lives below presentation, in API-serialization /
// view-model code, never inside a shared component). This module exists
// as the SECOND, defensive layer the CD3-D design calls for: a fail-closed
// parser for whatever an API response actually contains, so a component
// never has to re-derive authorization itself and never trusts an
// untyped/legacy value it wasn't given real authorization for.
//
// Components (ScoreDisplay, ConfidenceBadge, direction badges, graph
// edges, verdict components) become dumb renderers: they receive an
// AuthorizedClaim-shaped prop and render UNAVAILABLE as an explicit
// unavailable state -- never `null -> 0%`, `unknown -> "Neutral"`,
// `missing relationship -> "causes"`. Those defaults are exactly the
// regression category this phase exists to close.

export type Capability =
  | "observed_direction"
  | "historical_description"
  | "analytical_hypothesis"
  | "causal_relationship"
  | "evidence_quality"
  | "forecast"
  | "recommendation";

export type Strength = "authorized" | "qualified" | "unavailable";

export interface AuthorizedClaim {
  capability: Capability;
  strength: Strength;
  reason?: string | null;
}

// Mirrors app.services.claim_provenance.ClaimProvenance /
// RippleEvidenceState and app.services.measurement_semantics.
// IntegrityStatus (no TS type for these existed before CD3-D; frontend
// CD3-B consumers used inline string literals matching the backend
// values directly). Kept in sync by inspection, same convention as
// every other Python<->TS mirror in this codebase.
export type ClaimProvenanceValue =
  | "price_sign" | "event_direction" | "analytical_hypothesis"
  | "historical_outcome" | "fallback" | "unavailable" | "unknown";
export type RippleEvidenceStateValue = "observed" | "supported" | "hypothesized" | "unavailable";
export type IntegrityStatusValue = "valid" | "degraded" | "fallback" | "unavailable" | "invalid";

// ── Authorization decision logic — mirrors claim_authorization.py's
// three functions exactly. This IS enforcement logic living in the
// frontend, which looks like it contradicts "authorization lives below
// presentation" -- it doesn't: this is not a component deciding for
// itself, it's the same SHARED, reusable, non-bypassable-by-a-new-page
// contract as the Python version, callable from view-model code before
// any component ever sees the data (server components building page
// props, not JSX). The two implementations must be changed together;
// test_claim_authorization.py (backend) and claimAuthorization.test.ts
// (frontend) both pin the exact same decision table so drift is caught
// immediately, not silently. ────────────────────────────────────────

export function authorizeDirection(
  provenance: ClaimProvenanceValue,
  integrity: IntegrityStatusValue = "valid",
): AuthorizedClaim {
  if (integrity !== "valid") {
    return { capability: "analytical_hypothesis", strength: "unavailable", reason: `integrity_status=${integrity}` };
  }
  if (provenance === "price_sign") return { capability: "observed_direction", strength: "authorized" };
  if (provenance === "historical_outcome") return { capability: "historical_description", strength: "authorized" };
  if (provenance === "analytical_hypothesis" || provenance === "event_direction") {
    return { capability: "analytical_hypothesis", strength: "qualified" };
  }
  return { capability: "analytical_hypothesis", strength: "unavailable", reason: `provenance=${provenance}` };
}

export function authorizeRipple(
  evidenceState: RippleEvidenceStateValue,
  integrity: IntegrityStatusValue = "valid",
): AuthorizedClaim {
  if (integrity !== "valid") {
    return { capability: "causal_relationship", strength: "unavailable", reason: `integrity_status=${integrity}` };
  }
  if (evidenceState === "observed") return { capability: "causal_relationship", strength: "authorized" };
  if (evidenceState === "supported") return { capability: "causal_relationship", strength: "qualified" };
  if (evidenceState === "hypothesized") return { capability: "causal_relationship", strength: "qualified" };
  return { capability: "causal_relationship", strength: "unavailable", reason: `evidence_state=${evidenceState}` };
}

export function authorizeMeasurement(
  measurementType: string,
  integrity: IntegrityStatusValue,
): AuthorizedClaim {
  if (integrity !== "valid") {
    return { capability: "evidence_quality", strength: "unavailable", reason: `integrity_status=${integrity}` };
  }
  if (measurementType === "self_reported_certainty") {
    return { capability: "evidence_quality", strength: "qualified", reason: "self-reported, not independently verified" };
  }
  if (["evidence_composite", "hybrid_rubric", "deterministic_metric", "historical_calibration"].includes(measurementType)) {
    return { capability: "evidence_quality", strength: "authorized" };
  }
  return { capability: "evidence_quality", strength: "unavailable", reason: `measurement_type=${measurementType}` };
}

// FORECAST/RECOMMENDATION: no authorize_* function ever produces these
// capabilities. Named explicitly, permanently unavailable -- same
// reasoning as the Python constants.
export const FORECAST_UNAVAILABLE: AuthorizedClaim = {
  capability: "forecast", strength: "unavailable",
  reason: "No producer in the pipeline generates a verified forward-looking claim today",
};
export const RECOMMENDATION_UNAVAILABLE: AuthorizedClaim = {
  capability: "recommendation", strength: "unavailable",
  reason: "MarketRipple does not issue investment instructions",
};

const KNOWN_CAPABILITIES: Capability[] = [
  "observed_direction", "historical_description", "analytical_hypothesis",
  "causal_relationship", "evidence_quality", "forecast", "recommendation",
];
const KNOWN_STRENGTHS: Strength[] = ["authorized", "qualified", "unavailable"];

/** Fail-safe parse of a capability string from an API response --
 * unrecognized/missing values are NOT assumed; callers must treat a
 * `null` return the same as UNAVAILABLE (there is no "unknown but
 * renderable" state). */
export function parseCapability(raw: string | null | undefined): Capability | null {
  return (KNOWN_CAPABILITIES as string[]).includes(raw ?? "") ? (raw as Capability) : null;
}

/** Fail-safe parse of a strength string -- a missing/unrecognized value
 * resolves to "unavailable", the weakest state, never "authorized". */
export function parseStrength(raw: string | null | undefined): Strength {
  return (KNOWN_STRENGTHS as string[]).includes(raw ?? "") ? (raw as Strength) : "unavailable";
}

/**
 * The one gate every consumer should check before rendering ANYTHING
 * from an AuthorizedClaim -- false for "unavailable" strength, false for
 * a missing/malformed claim, false for an unrecognized capability. This
 * mirrors claim_authorization.py's own fail-closed contract: a claim
 * this function rejects must render an explicit unavailable state, not
 * a fabricated default.
 */
export function isRenderable(claim: AuthorizedClaim | null | undefined): boolean {
  if (!claim) return false;
  if (parseStrength(claim.strength) === "unavailable") return false;
  if (parseCapability(claim.capability) === null) return false;
  return true;
}

/**
 * Parses a raw, possibly-legacy/untyped value from an API response into
 * a safe AuthorizedClaim. Use this at the view-model boundary for any
 * endpoint that hasn't been migrated to emit real AuthorizedClaim JSON
 * yet -- it NEVER upgrades an untyped value into "authorized"; anything
 * it can't positively identify resolves to unavailable.
 */
export function coerceAuthorizedClaim(raw: unknown): AuthorizedClaim {
  if (!raw || typeof raw !== "object") {
    return { capability: "evidence_quality", strength: "unavailable", reason: "no claim data" };
  }
  const r = raw as Record<string, unknown>;
  const capability = parseCapability(typeof r.capability === "string" ? r.capability : null);
  if (!capability) {
    return { capability: "evidence_quality", strength: "unavailable", reason: "unrecognized capability" };
  }
  return {
    capability,
    strength: parseStrength(typeof r.strength === "string" ? r.strength : null),
    reason: typeof r.reason === "string" ? r.reason : null,
  };
}

File: apps/backend/app/services/claim_authorization.py
"""
CD3-D — Central Public Claim Authorization.

CD3-B (claim_provenance.py) answers "what kind of signal produced this
directional/relationship value." CD3-C (measurement_semantics.py) answers
"what kind of number is this, and is this specific instance trustworthy."
Neither enforces anything -- both were explicitly scoped to tagging data,
not deciding what a consumer may say. The CD3-D audit found exactly the
predicted consequence: dozens of independent consumers each decided for
themselves what to render, several silently discarding the B/C tags
along the way and reconstructing a stronger claim than the data
supports (an event-company's real, mixed impact_type collapsed to a
hardcoded "positive" in page.tsx; a HYPOTHESIZED Ripple edge rendered as
an unlabeled "-> causes" arrow, defaulting to "causes" even when the
relationship was undefined; a homepage template sentence built a
forecast ("will likely be led by X") with no forecast-capable producer
anywhere in the pipeline).

This module is the single place that answers the owner's own framing of
the question: **"Given this claim's type, provenance, evidence state,
measurement semantics, and integrity status, what is MarketRipple
allowed to tell the user?"**

Deliberately NOT a giant `isClaimAllowed()` enumerating page-specific
rules. A small, reusable CAPABILITY model instead: every public claim is
an attempt to exercise one of 7 capabilities (see `Capability` below).
A claim's real provenance/evidence-state/measurement-type, combined with
its current integrity_status, determines whether that capability is
granted -- and at what strength (AUTHORIZED / QUALIFIED / UNAVAILABLE).
Two capabilities (FORECAST, RECOMMENDATION) have NO legitimate producer
anywhere in the current pipeline and resolve UNAVAILABLE unconditionally
-- this is not a gap to be "fixed" by inventing complicated authorization
rules to rescue existing forecast/recommendation-shaped output. If a
future producer earns one of these (e.g. a real calibrated, benchmark-
relative forecast engine), it gets a real new authorization rule then --
not before.

Enforcement location, per the owner's explicit correction when opening
this phase: authorization must live BELOW presentation, not inside a
shared component (a component-level gate is one new caller away from
being bypassed). This module is imported by API-serialization / view-
model code (D2 onward), never by a React component directly -- shared
components (D7) become dumb renderers that trust the AuthorizedClaim
they're handed, they never re-derive authorization themselves.

Fail-closed contract, matching claim_provenance.py/measurement_
semantics.py exactly: UNKNOWN provenance/measurement-type, and any
non-VALID integrity_status, never yield AUTHORIZED. A missing/legacy
value must never be inferred into a stronger claim than the data proves.
"""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from app.services.claim_provenance import ClaimProvenance, RippleEvidenceState
from app.services.measurement_semantics import IntegrityStatus, MeasurementType


class Capability(str, Enum):
    """What kind of public claim is being attempted. A claim's real
    provenance/evidence-state/measurement-type determines which single
    capability is even relevant to it (see the `authorize_*` functions);
    FORECAST and RECOMMENDATION are listed for completeness -- and
    because their permanent UNAVAILABLE status is itself part of the
    contract -- even though no `authorize_*` function currently grants
    either."""

    # A real, already-observed price/index/flow movement, stated as fact.
    OBSERVED_DIRECTION = "observed_direction"
    # A real, measured past outcome, described in past tense only.
    HISTORICAL_DESCRIPTION = "historical_description"
    # An LLM's own analytical judgment about a specific entity/sector.
    ANALYTICAL_HYPOTHESIS = "analytical_hypothesis"
    # A claim that one thing produced/affects another (Ripple/graph edges).
    CAUSAL_RELATIONSHIP = "causal_relationship"
    # The right to state how confident/well-evidenced a claim is.
    EVIDENCE_QUALITY = "evidence_quality"
    # A genuinely forward-looking statement about what will happen.
    # No current producer earns this -- see FORECAST_UNAVAILABLE below.
    FORECAST = "forecast"
    # An instruction to act (buy/sell/hold/allocate).
    # No current producer earns this -- see RECOMMENDATION_UNAVAILABLE.
    RECOMMENDATION = "recommendation"


class Strength(str, Enum):
    """How strongly a granted capability may be exercised."""

    # The claim may be stated at the producer's own real strength.
    AUTHORIZED = "authorized"
    # The claim may be stated, but only in visibly hedged/qualified
    # wording -- never a bare declarative badge/label.
    QUALIFIED = "qualified"
    # Nothing may be publicly claimed. Render an explicit unavailable
    # state, never infer a fallback value (never "neutral", never 0%,
    # never "causes" as a default relationship).
    UNAVAILABLE = "unavailable"


@dataclass(frozen=True)
class AuthorizedClaim:
    """The result of an authorization decision. `reason` is set whenever
    strength is QUALIFIED or UNAVAILABLE, explaining why -- consumers
    may show it as a tooltip/disclosure, matching the pattern CD3-C's
    fixes already established (e.g. "the model's own self-rated
    certainty, not a verified score")."""

    capability: Capability
    strength: Strength
    reason: str | None = None


def _claim(capability: Capability, strength: Strength, reason: str | None = None) -> AuthorizedClaim:
    return AuthorizedClaim(capability=capability, strength=strength, reason=reason)


# ── Directional claims (ClaimProvenance) ─────────────────────────────────────

def authorize_direction(
    provenance: ClaimProvenance,
    integrity: IntegrityStatus = IntegrityStatus.VALID,
) -> AuthorizedClaim:
    """Authorizes a directional/impact claim (e.g. "RELIANCE fell 0.5%",
    "Potential positive impact on RELIANCE"). Any non-VALID integrity
    status collapses straight to UNAVAILABLE regardless of provenance --
    a real self-reported hypothesis that's currently DEGRADED/FALLBACK/
    INVALID authorizes nothing, matching the owner's explicit rule that
    FALLBACK/DEGRADED/INVALID always resolve UNAVAILABLE."""
    if integrity != IntegrityStatus.VALID:
        return _claim(Capability.ANALYTICAL_HYPOTHESIS, Strength.UNAVAILABLE,
                       reason=f"integrity_status={integrity.value}")

    if provenance == ClaimProvenance.PRICE_SIGN:
        return _claim(Capability.OBSERVED_DIRECTION, Strength.AUTHORIZED)
    if provenance == ClaimProvenance.HISTORICAL_OUTCOME:
        return _claim(Capability.HISTORICAL_DESCRIPTION, Strength.AUTHORIZED)
    if provenance in (ClaimProvenance.ANALYTICAL_HYPOTHESIS, ClaimProvenance.EVENT_DIRECTION):
        # EVENT_DIRECTION is an event-level LLM read broadcast to every
        # matched company -- never stronger than the sector/entity-level
        # hypothesis it actually is, so it gets the same capability/
        # strength as a direct analytical hypothesis, never AUTHORIZED.
        return _claim(Capability.ANALYTICAL_HYPOTHESIS, Strength.QUALIFIED)
    # FALLBACK, UNAVAILABLE, UNKNOWN -- no legitimate directional claim.
    return _claim(Capability.ANALYTICAL_HYPOTHESIS, Strength.UNAVAILABLE,
                   reason=f"provenance={provenance.value}")


# ── Ripple / relationship claims (RippleEvidenceState) ───────────────────────

def authorize_ripple(
    evidence_state: RippleEvidenceState,
    integrity: IntegrityStatus = IntegrityStatus.VALID,
) -> AuthorizedClaim:
    """Authorizes a causal/relationship claim (a Ripple edge, an
    Intelligence Graph edge, Deep Research second-order effects). Per
    the CD3-A audit, 100% of live Ripple/graph data sampled was
    HYPOTHESIZED -- this function returns QUALIFIED for that state, not
    AUTHORIZED, exactly matching the owner's explicit example ("Possible
    transmission mechanism", never "X causes Y"). An UNDEFINED/missing
    evidence_state must resolve here to UNAVAILABLE -- never silently to
    HYPOTHESIZED or any stronger state; callers are responsible for
    passing RippleEvidenceState.UNAVAILABLE (via claim_provenance.py's
    own get_ripple_evidence_state() fail-safe accessor) rather than None."""
    if integrity != IntegrityStatus.VALID:
        return _claim(Capability.CAUSAL_RELATIONSHIP, Strength.UNAVAILABLE,
                       reason=f"integrity_status={integrity.value}")

    if evidence_state == RippleEvidenceState.OBSERVED:
        return _claim(Capability.CAUSAL_RELATIONSHIP, Strength.AUTHORIZED)
    if evidence_state == RippleEvidenceState.SUPPORTED:
        return _claim(Capability.CAUSAL_RELATIONSHIP, Strength.QUALIFIED)
    if evidence_state == RippleEvidenceState.HYPOTHESIZED:
        return _claim(Capability.CAUSAL_RELATIONSHIP, Strength.QUALIFIED)
    # UNAVAILABLE, UNKNOWN, or anything else -- no relationship assertion.
    return _claim(Capability.CAUSAL_RELATIONSHIP, Strength.UNAVAILABLE,
                   reason=f"evidence_state={getattr(evidence_state, 'value', evidence_state)}")


# ── Measurement / confidence claims (MeasurementType) ────────────────────────

def authorize_measurement(
    measurement_type: MeasurementType,
    integrity: IntegrityStatus,
) -> AuthorizedClaim:
    """Authorizes the right to state how confident/well-evidenced a
    claim is. SELF_REPORTED_CERTAINTY is always QUALIFIED, never
    AUTHORIZED -- it may be shown, but only with its self-report nature
    disclosed (matching every CD3-C fix's wording pattern), never as a
    bare, undifferentiated percentage. Real composites/deterministic
    metrics/calibration are AUTHORIZED at their own real scale -- callers
    must still use the real unit/label (e.g. "Evidence Coverage", not a
    bare "%"), this function only answers whether disclosure is
    required, not what the disclosure says."""
    if integrity != IntegrityStatus.VALID:
        return _claim(Capability.EVIDENCE_QUALITY, Strength.UNAVAILABLE,
                       reason=f"integrity_status={integrity.value}")

    if measurement_type == MeasurementType.SELF_REPORTED_CERTAINTY:
        return _claim(Capability.EVIDENCE_QUALITY, Strength.QUALIFIED,
                       reason="self-reported, not independently verified")
    if measurement_type in (
        MeasurementType.EVIDENCE_COMPOSITE, MeasurementType.HYBRID_RUBRIC,
        MeasurementType.DETERMINISTIC_METRIC, MeasurementType.HISTORICAL_CALIBRATION,
    ):
        return _claim(Capability.EVIDENCE_QUALITY, Strength.AUTHORIZED)
    # DERIVED_TRANSFORM, UNKNOWN -- no claim about evidence quality here;
    # a derived transform's own inputs must be authorized by their own
    # producers, this function can't see them (matches measurement_
    # semantics.is_publicly_authorized's identical DERIVED_TRANSFORM rule).
    return _claim(Capability.EVIDENCE_QUALITY, Strength.UNAVAILABLE,
                   reason=f"measurement_type={measurement_type.value}")


# ── Capabilities with no legitimate producer today ───────────────────────────
# Not omitted from the model -- named explicitly, permanently UNAVAILABLE,
# so a future implementer sees the boundary rather than inventing ad hoc
# authorization logic to make an existing forecast/recommendation-shaped
# UI element "fit." RECOMMENDATION is also independently backstopped by
# recommendation_language.py's generation-time blacklist; FORECAST is why
# historical_forecast_guard.py exists. Both stay UNAVAILABLE until a real
# producer for either is built and reviewed on its own merits.

FORECAST_UNAVAILABLE = AuthorizedClaim(
    capability=Capability.FORECAST, strength=Strength.UNAVAILABLE,
    reason="No producer in the pipeline generates a verified forward-looking claim today",
)
RECOMMENDATION_UNAVAILABLE = AuthorizedClaim(
    capability=Capability.RECOMMENDATION, strength=Strength.UNAVAILABLE,
    reason="MarketRipple does not issue investment instructions",
)
